R/ard_stack_hierarchical.R

```diff
@@ -0,0 +1,460 @@
+  if (!is_empty(id)) data <- dplyr::arrange(data, dplyr::pick(all_of(c(id, by, variables)))) # styler: off
+
+  # print denom columns if not 100% clear which are used
+  if (!is_empty(id) && is.data.frame(denominator)) {
+    denom_cols <- intersect(by, names(denominator))
+    if (!setequal(by, denom_cols)) {
+      msg <-
+        ifelse(
+          is_empty(denom_cols),
+          "Denominator set by number of rows in {.arg denominator} data frame.",
+          "Denominator set by {.val {denom_cols}} column{?s} in {.arg denominator} data frame."
+        )
+      cli::cli_inform(c("i" = msg))
+    }
+  }
+
+  # go about calculating the statistics within the variables -------------------
+  # define index in `variables` that also appear in `include`
+  which_include <- which(variables %in% include)
+
+  lst_results <- list()
+  for (i in which_include) {
+    lst_results <-
+      lst_results |>
+      append(
+        .run_hierarchical_fun(
+          data = data,
+          variables = variables[seq_len(i)],
+          by = by,
+          denominator = denominator,
+          id = id
+        ) |>
+          list()
+      )
+  }
+
+  # calculate results overall if requested -------------------------------------
+  if (isTRUE(overall)) {
+    for (i in which_include) {
+      lst_results <-
+        lst_results |>
+        append(
+          .run_hierarchical_fun(
+            data = data,
+            variables = variables[seq_len(i)],
+            by = setdiff(by, names(denominator)),
+            denominator = denominator,
+            id = id
+          ) |>
+            list()
+        )
+
+      # if there are columns in `by` not present in `denominator`, re-run with `by = NULL`
+      if (!is_empty(setdiff(by, names(denominator)))) {
+        lst_results <-
+          lst_results |>
+          append(
+            .run_hierarchical_fun(
+              data = data,
+              variables = variables[seq_len(i)],
+              by = NULL,
+              denominator = denominator,
+              id = id
+            ) |>
+              list()
+          )
+      }
+    }
+  }
+
+  # add univariate tabulations of by variables ---------------------------------
+  if (is.data.frame(denominator) && !is_empty(intersect(by, names(denominator)))) {
+    lst_results <-
+      lst_results |>
+      append(
+        ard_categorical(
+          data = denominator,
+          variables = all_of(intersect(by, names(denominator)))
+        ) |>
+          list()
+      )
+  }
+
+  # add overall row if requested -----------------------------------------------
+  if (isTRUE(overall_row)) {
+    lst_results <-
+      lst_results |>
+      append(
+        .run_hierarchical_fun(
+          data = data |> dplyr::mutate(..ard_hierarchical_overall.. = TRUE),
+          variables = "..ard_hierarchical_overall..",
+          by = by,
+          denominator = denominator,
+          id = id
+        ) |>
```

Comment:
I think I need to run the original function here with `overall_row = FALSE`, so we get the `overall=TRUE` results for this if they were requested?

Reply:
has now been added